contracts/src/v0.8/shared/test/testhelpers/GasConsumer.sol

```diff
@@ -0,0 +1,26 @@
+pragma solidity ^0.8.0;
+
+contract GasConsumer {
+  function consumeAllGas() external view {
+    assembly {
+      // While loop that operates indefinitely, written in yul to ensure better granularity over exactly how much gas is spent
+      for {
+        // Loop will run forever since 0 < 1
+      } lt(0, 1) {
+
+      } {
+        // If 30 gas is remaining, then exit the loop by returning
+        if lt(gas(), 30) {
+          pop(add(0, 0)) // Add two numbers but don't push result onto the stack. Safely consume any residual gas from an odd number remaining.
```

Comment:
good job on going granular, please explain how the values are calculated, the important thing is the `why`, e.g:
1. why terminate at 30, but not like 40, or 20
2. why consume residual gas via `pop(add(0, 0))` as opposed more loop iterations

Reply:
That value was calculated via manual binary search to determine the least amount of gas possible which can be returned. 30 was the least amount of gas that could remain before an OOG error would be thrown on this call.

Based on what I was able to deduce, one more iteration of the loop would cost more gas than remains, but you need to consume the remaining gas. I.E you leave 30 gas because 1 more loop would cost >30, but you still need to consume the gas. Add is one of the cheapest non-state-changing opcode you can perform, but yul requires you to pop it off the stack of assign it to a variable, and using `pop` ensures we're not charged the extraneous cost for pushing onto the stack. I will add all of these to comments.

Reply:
this may not be reliable right? for example if you need 20 gas to return, with current setup it might reach `gas()` with 28 gas, and it is able to return, however if someone changes the gas limit passed in, and if it happens to be 22 gas remaining when reaching `gas()`, then the test might fail.

Reply:
maybe consider a range of input gases, eg. 20k, 20k+1, 20k+2,... 20k+30
and locate the highest gas cutoff number that make all of them pass